pkg/controller/networkpolicy/controller_querier_replier.go

```diff
@@ -0,0 +1,183 @@
+	PolicyRef
+	selector metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+type Rule struct {
+	PolicyRef
+	Direction networkingv1beta1.Direction `json:"direction,omitempty"`
+	RuleIndex int                         `json:"ruleindex,omitempty"`
+}
+
+// NewNetworkPolicyController returns a new *NetworkPolicyController.
+func NewEndpointQueryReplier(networkPolicyController *NetworkPolicyController) *EndpointQueryReplier {
+	n := &EndpointQueryReplier{
+		networkPolicyController: networkPolicyController,
+	}
+	return n
+}
+
+// QueryNetworkPolicies returns kubernetes network policy references relevant to the selected network endpoint. Relevant
+// policies fall into three categories: applied policies (Policies in Endpoint type) are policies which directly apply to
+// an endpoint, egress and ingress rules (Rules in Endpoint type) are policies which reference the endpoint in an ingress/
+// egress rule respectively.
+func (eq EndpointQueryReplier) QueryNetworkPolicies(namespace string, podName string) (*EndpointQueryResponse, error) {
+	// check if namespace and podName select an existing pod
+	_, err := eq.networkPolicyController.podInformer.Lister().Pods(namespace).Get(podName)
+	if err != nil {
+		return nil, nil
+	}
+	type ruleTemp struct {
+		policy *antreatypes.NetworkPolicy
+		index  int
+	}
+	// create network policies categories
+	applied := make([]*antreatypes.NetworkPolicy, 0)
+	ingress := make([]*ruleTemp, 0)
+	egress := make([]*ruleTemp, 0)
+	// get all appliedToGroups using pod index, then get applied policies using appliedToGroup
+	appliedToGroups, err := eq.networkPolicyController.appliedToGroupStore.GetByIndex(store.PodIndex, podName+"/"+namespace)
+	if err != nil {
+		return nil, err
+	}
+	for _, appliedToGroup := range appliedToGroups {
+		policies, err := eq.networkPolicyController.internalNetworkPolicyStore.GetByIndex(store.AppliedToGroupIndex,
+			string(appliedToGroup.(*antreatypes.AppliedToGroup).UID))
+		if err != nil {
+			return nil, err
+		}
+		for _, policy := range policies {
+			applied = append(applied, policy.(*antreatypes.NetworkPolicy))
+		}
+	}
+	// get all addressGroups using pod index, then get ingress and egress policies using addressGroup
+	addressGroups, err := eq.networkPolicyController.addressGroupStore.GetByIndex(store.PodIndex, podName+"/"+namespace)
+	if err != nil {
+		return nil, err
+	}
+	for _, addressGroup := range addressGroups {
+		policies, err := eq.networkPolicyController.internalNetworkPolicyStore.GetByIndex(store.AddressGroupIndex,
+			string(addressGroup.(*antreatypes.AddressGroup).UID))
+		if err != nil {
+			return nil, err
+		}
+		for _, policy := range policies {
+			for i, rule := range policy.(*antreatypes.NetworkPolicy).Rules {
+				for _, addressGroupTrial := range rule.To.AddressGroups {
+					if addressGroupTrial == string(addressGroup.(*antreatypes.AddressGroup).UID) {
+						egress = append(egress, &ruleTemp{policy: policy.(*antreatypes.NetworkPolicy), index: i})
```

Comment:
nit: consider breaking out of the loop here. Same for ingress case.

Reply:
I suppose that is a good idea. Is there ever a case where a policy references an endpoint in both an ingress and egress rule?

Reply:
There will be cases where a policy references an endpoint in both an ingress and egress rule. I'm not talking about breaking out of the outer loop; for the inner loop, once we find out an addressGroup is referred to in an ingress/egress rule, we can break out of the loop as soon as we the group matches without checking the other addressGroups. This should be a very minor optimization though, but could be beneficial if the addressGroups in a rule is a long list.

Reply:
Thats sounds good to me.

Reply:
Fixed in https://github.com/vmware-tanzu/antrea/pull/1116